arangod/Aql/HashedCollectExecutor.cpp

```diff
@@ -0,0 +1,299 @@
+
+  if (_infos.getAggregateVariables().empty()) {
+    // no aggregate registers. simply increase the counter
+    if (_infos.getCount()) {
+      // TODO get rid of this special case if possible
+      TRI_ASSERT(!aggregateValues->empty());
+      aggregateValues->back()->reduce(EmptyValue);
+    }
+  } else {
+    // apply the aggregators for the group
+    TRI_ASSERT(aggregateValues->size() == _infos.getAggregatedRegisters().size());
+    size_t j = 0;
+    for (auto const& r : _infos.getAggregatedRegisters()) {
+      if (r.second == ExecutionNode::MaxRegisterId) {
+        (*aggregateValues)[j]->reduce(EmptyValue);
+      } else {
+        (*aggregateValues)[j]->reduce(input.getValue(r.second));
+      }
+      ++j;
+    }
+  }
+}
+
+void HashedCollectExecutor::writeCurrentGroupToOutput(OutputAqlItemRow& output) {
+  // build the result
+  TRI_ASSERT(!_infos.getCount() || _infos.getCollectRegister() != ExecutionNode::MaxRegisterId);
+
+  auto& keys = _currentGroup->first;
+  TRI_ASSERT(_currentGroup->second != nullptr);
+
+  TRI_ASSERT(keys.size() == _infos.getGroupRegisters().size());
+  size_t i = 0;
+  for (auto& key : keys) {
+    output.cloneValueInto(_infos.getGroupRegisters()[i++].first,
+                          _lastInitializedInputRow, key);
+    const_cast<AqlValue*>(&key)->erase();  // to prevent double-freeing later
+  }
+
+  if (!_infos.getCount()) {
+    TRI_ASSERT(_currentGroup->second->size() == _infos.getAggregatedRegisters().size());
+    size_t j = 0;
+    for (auto const& r : *(_currentGroup->second)) {
+      output.cloneValueInto(_infos.getAggregatedRegisters()[j++].first,
+                            _lastInitializedInputRow, r->stealValue());
+    }
+  } else {
+    // set group count in result register
+    TRI_ASSERT(!_currentGroup->second->empty());
+    output.cloneValueInto(_infos.getCollectRegister(), _lastInitializedInputRow,
+                          _currentGroup->second->back()->stealValue());
+  }
+}
+
+ExecutionState HashedCollectExecutor::init() {
+  TRI_ASSERT(!_isInitialized);
+
+  // fetch & consume all input
+  while (_upstreamState != ExecutionState::DONE) {
+    InputAqlItemRow input = InputAqlItemRow{CreateInvalidInputRowHint{}};
+    std::tie(_upstreamState, input) = _fetcher.fetchRow();
+
+    if (_upstreamState == ExecutionState::WAITING) {
+      TRI_ASSERT(!input.isInitialized());
+      return ExecutionState::WAITING;
+    }
+
+    // !input.isInitialized() => _upstreamState == ExecutionState::DONE
+    TRI_ASSERT(input.isInitialized() || _upstreamState == ExecutionState::DONE);
+
+    // needed to remember the last valid input aql item row
+    // NOTE: this might impact the performance
+    if (input.isInitialized()) {
+      _lastInitializedInputRow = input;
+
+      consumeInputRow(input);
+    }
+  }
+
+  // initialize group iterator for output
+  if (!_producingOutput) {
+    _producingOutput = true;
+    // setup the iterator
+    _currentGroup = _allGroups.begin();
+  }
```

Comment:
`_producingOutput` and the `if` can be removed now that this is moved to `init`.